sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,1085 @@
+}
+
+// copied from crypto/ocsp
+func getOIDFromHashAlgorithm(target crypto.Hash) (asn1.ObjectIdentifier, error) {
+	for hash, oid := range hashOIDs {
+		if hash == target {
+			return oid, nil
+		}
+	}
+	return nil, fmt.Errorf("no valid OID is found for the hash algorithm: %v", target)
+}
+
+func (c *Client) ClearCache() {
+	c.ocspResponseCache.Purge()
+}
+
+func (c *Client) getHashAlgorithmFromOID(target pkix.AlgorithmIdentifier) crypto.Hash {
+	for hash, oid := range hashOIDs {
+		if oid.Equal(target.Algorithm) {
+			return hash
+		}
+	}
+	// no valid hash algorithm is found for the oid. Falling back to SHA1
+	return crypto.SHA1
+}
+
+// isInValidityRange checks the validity
+func isInValidityRange(currTime, nextUpdate time.Time) bool {
+	return !nextUpdate.IsZero() && !currTime.After(nextUpdate)
+}
+
+func extractCertIDKeyFromRequest(ocspReq []byte) (*certIDKey, *ocspStatus) {
+	r, err := ocsp.ParseRequest(ocspReq)
+	if err != nil {
+		return nil, &ocspStatus{
+			code: ocspFailedDecomposeRequest,
+			err:  err,
+		}
+	}
+
+	// encode CertID, used as a key in the cache
+	encodedCertID := &certIDKey{
+		r.HashAlgorithm,
+		base64.StdEncoding.EncodeToString(r.IssuerNameHash),
+		base64.StdEncoding.EncodeToString(r.IssuerKeyHash),
+		r.SerialNumber.String(),
+	}
+	return encodedCertID, &ocspStatus{
+		code: ocspSuccess,
+	}
+}
+
+func (c *Client) encodeCertIDKey(certIDKeyBase64 string) (*certIDKey, error) {
+	r, err := base64.StdEncoding.DecodeString(certIDKeyBase64)
+	if err != nil {
+		return nil, err
+	}
+	var cid certID
+	rest, err := asn1.Unmarshal(r, &cid)
+	if err != nil {
+		// error in parsing
+		return nil, err
+	}
+	if len(rest) > 0 {
+		// extra bytes to the end
+		return nil, err
+	}
+	return &certIDKey{
+		c.getHashAlgorithmFromOID(cid.HashAlgorithm),
+		base64.StdEncoding.EncodeToString(cid.NameHash),
+		base64.StdEncoding.EncodeToString(cid.IssuerKeyHash),
+		cid.SerialNumber.String(),
+	}, nil
+}
+
+func decodeCertIDKey(k *certIDKey) (string, error) {
+	serialNumber := new(big.Int)
+	serialNumber.SetString(k.SerialNumber, 10)
+	nameHash, err := base64.StdEncoding.DecodeString(k.NameHash)
+	if err != nil {
+		return "", err
+	}
+	issuerKeyHash, err := base64.StdEncoding.DecodeString(k.IssuerKeyHash)
+	if err != nil {
+		return "", err
+	}
+	hashAlgoOid, err := getOIDFromHashAlgorithm(k.HashAlgorithm)
+	if err != nil {
+		return "", err
+	}
+	encodedCertID, err := asn1.Marshal(certID{
+		pkix.AlgorithmIdentifier{
+			Algorithm:  hashAlgoOid,
+			Parameters: asn1.RawValue{Tag: 5 /* ASN.1 NULL */},
+		},
+		nameHash,
+		issuerKeyHash,
+		serialNumber,
+	})
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(encodedCertID), nil
+}
+
+func (c *Client) checkOCSPResponseCache(encodedCertID *certIDKey, subject, issuer *x509.Certificate) (*ocspStatus, error) {
+	c.ocspResponseCacheLock.RLock()
+	var cacheValue *ocspCachedResponse
+	v, ok := c.ocspResponseCache.Get(*encodedCertID)
```

Comment:
@sgmiller The return value is correctly a pointer, but the key here is a fat struct.